Docs 4430 crypto examples v4.2 (#945)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/cryptography-jce.adoc

You can use cryptography capabilities in two ways:

> * Password-based encryption (PBE): Encrypt/sign content by providing only an encryption password.

Reviewer: GLOBALLY replace / with the better of "and" or "or."

Reviewer: For parallelism with next list item:
```suggestion
* Password-based encryption (PBE)
This method enables you to encrypt/sign content by providing only an encryption password.
```

> * Key-based Encryption: Similar to how PGP and XML encryption works, this method enables you to configure a symmetric or asymmetric key to perform encryption/signing operations.

Reviewer: ```suggestion
* Key-based encryption
Similar to how PGP and XML encryption works, this method enables you to configure a symmetric or asymmetric key to perform encryption/signing operations.
```
DITTO re. /

You can encrypt all, or part of a message using any of these two methods.

== PBE

This method applies a hash function over the provided password to generate a symmetric key that is compatible with standard encryption algorithms. Because PBE only requires a password, a global configuration element is not needed for the PBE operations.

> === Configure Password-Based Encryption From Anypoint Studio

Reviewer: ```suggestion
=== Configure Password-Based Encryption from Anypoint Studio
```

> === Configuring Password-Based Encryption From Anypoint Studio

Reviewer 2: Configuring > Configure

To configure PBE from Anypoint Studio, follow these steps:

> . Add the Crypto module to your project.

Reviewer 2: From the Mule Palette, add the Crypto module to your project.

> . From the Mule palette, add the Crypto module to your project.

Reviewer: I performed a brief internet search and saw only Cryptography module; should we spell this out?

+
See xref:cryptography.adoc#install-crypto-module[Install the Extension] for instructions.
. Select the desired operation, and drag the component to the flow:
+
image::mruntime-crypto-pbe-add.png[crypto-pbe-add]

> . Configure the *Algorithm* and *Password* properties:

Reviewer 2: In the JCE encrypt pbe component view, configure the *Algorithm* and *Password* properties:

+
image::mruntime-crypto-pbe-config.png[crypto-pbe-config]

=== XML Examples

The following are XML examples for each each of the PBE operations:

> * Example: PBE Encryption

Reviewer: I wonder if we really need "Example:" in the subtitles of an Examples container topic?
I suggest omitting.

+
[source,xml,linenums]
----
<crypto:jce-encrypt-pbe password="a-Sup3r_Secure-Passw0rd"/>
----
+
If no algorithm is specified, `PBEWithHmacSHA256AndAES_128` is used.

* Example: PBE Decryption
+
[source,xml,linenums]
----
<crypto:jce-decrypt-pbe algorithm="PBEWithHmacSHA256AndAES_128" password="a-Sup3r_Secure-Passw0rd"/>
----

* Example: PBE Signature
+
[source,xml,linenums]
----
<crypto:jce-sign-pbe password="a-Sup3r_Secure-Passw0rd"/>
----
+
If no algorithm is specified, `PBEWithHmacSHA256` is used.

* Example: PBE Signature Validation
+
[source,xml,linenums]
----
<crypto:jce-validate-pbe password="a-Sup3r_Secure-Passw0rd" algorithm="PBEWithHmacSHA256" expected="#[vars.expectedSignature]"/>
----
+
The `expected` parameter defines the signature used to validate the message.

== Key-Based Encryption

> Configure a symmetric or asymmetric key to perform encryption/signing operations.

Reviewer: DITTO re. /

> === Configuring Key-Based Encryption From Anypoint Studio

Reviewer 2: Configuring > Configure

> === Configure Key-Based Encryption From Anypoint Studio

Reviewer: ```suggestion
=== Configure Key-Based Encryption from Anypoint Studio
```

> To configure Key-based encryption operations from Anypoint Studio, follow these steps:

Reviewer: ```suggestion
To configure key-based encryption operations from Anypoint Studio, follow these steps:
```

> . From the Mule palette, add the Crypto module to your project.

Reviewer: DITTO re. Cryptography
If quoting the UI, consider...:
```suggestion
. From the Mule palette, add *Crypto* to your project.
```

+
See xref:cryptography.adoc#install-crypto-module[Install the Extension] for instructions.
. Select the desired operation, and drag the component to the flow:
+
image::mruntime-crypto-jce-add.png[crypto-jce-add]

> . Open the component properties and select an existing *Module configuration*, or create a new one by specifying the *Keystore*, *Type* (JKS, JCEKS, PKCS12) and *Password*.

Reviewer: ```suggestion
. Open the component properties and select an existing module configuration, or create a new one by specifying values for *Keystore*, *Type* (JKS, JCEKS, PKCS12), and *Password*.
```

+

> You can also add Symmetric or Asymmetric key info to be used in the sign operations:

Reviewer: ```suggestion
You can also add symmetric or asymmetric key information to be used in the sign operations:
```

Reviewer: These are lowercase in the text introducing the XML Examples topic, so...I would stick with lowercase unless quoting a UI label value.

+
image::mruntime-crypto-jce-global-config.png[crypto-jce-global-config]

> . Configure *Key selection* by using a *Key id* previously defined in the Module configuration, or define a new one for this operation:

Reviewer: ```suggestion
. Configure *Key selection* by using a *Key id* value previously defined in the module configuration, or define a new one for this operation:
```

+
image::mruntime-crypto-jce-config.png[crypto-jce-config]

> . Select the Algorithm to use during the operation.

Reviewer: ```suggestion
. Select the algorithm to use during the operation.
```

=== XML Examples

The following XML examples show a JCE configuration that defines symmetric and asymmetric keys and different operations using these keys.

> ==== Configuration

Reviewer: In general, we want to avoid nesting text down four or more levels.
When this happens, reevaluate topic organizations and see if it is possible to uplevel information.

> In this example, a keystore with different type of keys is defined in a JCE configuration:

Reviewer: ```suggestion
In this example, a keystore with different types of keys is defined in a JCE configuration:
```

</crypto:jce-config>
----

==== Asymmetric

> The following are examples of operations that use the asymmetric keys defined in the previous configuration:

Reviewer: ```suggestion
The following example operations use the asymmetric keys defined in the previous configuration.
```

> * Example: Asymmetric Encryption

Reviewer: DITTO re. "Example:" in the subtopic title

+
[source,xml,linenums]
----
<crypto:jce-encrypt config-ref="jceConfig" keyId="rsa" algorithm="RSA"/>
----

* Example: Asymmetric Decryption
+
[source,xml,linenums]
----
<crypto:jce-decrypt config-ref="jceConfig" keyId="rsa" algorithm="RSA"/>
----

==== Symmetric

> The following are examples of operations that use the symmetric keys defined in the previous configuration:

Reviewer: ```suggestion
The following example operations use the symmetric keys defined in the previous configuration.
```
